src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java

```diff
@@ -18,7 +27,282 @@
  */
 public class BlockModelBuilder extends ModelBuilder<BlockModelBuilder> {
 
-    public BlockModelBuilder(ResourceLocation outputLocation, ExistingFileHelper existingFileHelper) {
+    private static final Vector3f ONE = new Vector3f(1, 1, 1);
+
+    private final RootTransformBuilder rootTransform = new RootTransformBuilder();
+
+    public BlockModelBuilder(ResourceLocation outputLocation, ExistingFileHelper existingFileHelper)
+    {
         super(outputLocation, existingFileHelper);
     }
+
+    public RootTransformBuilder rootTransform()
+    {
+        return rootTransform;
+    }
+
+    @Override
+    public JsonObject toJson()
+    {
+        JsonObject json = super.toJson();
+
+        // If there were any transform properties set, add them to the output.
+        JsonObject transform = rootTransform.toJson();
+        if (transform.size() > 0)
+        {
+            json.add("transform", transform);
+        }
+
+        return json;
+    }
+
+    private JsonArray writeVec3(Vector3f vector)
+    {
+        JsonArray array = new JsonArray();
+        array.add(vector.x());
+        array.add(vector.y());
+        array.add(vector.z());
+        return array;
+    }
+
+    private JsonArray writeQuaternion(Quaternion quaternion)
+    {
+        JsonArray array = new JsonArray();
+        array.add(quaternion.i());
+        array.add(quaternion.j());
+        array.add(quaternion.k());
+        array.add(quaternion.r());
+        return array;
+    }
+
+    public class RootTransformBuilder
+    {
+        public enum Origin implements StringRepresentable
+        {
+            Center(new Vector3f(), "origin"),
+            Corner(new Vector3f(1f, 1f, 1f), "corner"),
+            OpposingCorner(new Vector3f(.5f, .5f, .5f), "opposing_corner");
+
+            private final Vector3f vec;
+            private final String name;
+
+            Origin(Vector3f vec, String name)
+            {
+                this.vec = vec;
+                this.name = name;
+            }
+
+            public Vector3f vec()
```

Comment:
This should be named in the style of a getter. Probably `getVector()` can suffice, unless there's a better name than just 'vector'.

Reply:
I've gone with getVector for now